### ISSUE TITLE
askForNotificationPermission is blocking the app

### ISSUE THREAD

**Issue by sbatezat** [Countly/countly-sdk-flutter-bridge#281]
I'm unable to successfully call the "Countly.askForNotificationPermission" method, just after the "Countly.initWithConfig" one. By debugging your dart code, I can see we've got a neverending call to the native code "askForNotificationPermission".

>     final String? result = await _channel.invokeMethod('askForNotificationPermission');

My tests have been done on Android.
A call to pushTokenType (on "TEST") has been successfully made before the init.

**Comment by ijunaid** [comment]
Hi @sbatezat,
Thank you for reaching out and providing details.

Which version of the Countly SDK are you currently using? Additionally, could you share the code you’re using to initialize the SDK, including the code for "askForNotificationPermission"?

From your description, it sounds like the Countly.askForNotificationPermission method is not returning as expected, resulting in a "never-ending" call. Is this correct? If possible, could you also specify any logs or error messages you’re seeing in the console?

This information will help us investigate the issue further. Looking forward to your response!

**Comment by sbatezat** [comment]
Hi,

> Which version of the Countly SDK are you currently using?

The latest stable, 24.7.1

> Additionally, could you share the code you’re using to initialize the SDK, including the code for "askForNotificationPermission"?

```
    await Countly.pushTokenType(Countly.messagingMode["TEST"]);

    final deviceId = await userRepository.getUserId();
    CountlyConfig config = CountlyConfig(countlyServer, countlyAppKey);
    config.enableCrashReporting();
    config.apm.enableAppStartTimeTracking().enableManualAppLoadedTrigger();
    config.apm.enableForegroundBackgroundTracking();
    if (deviceId != null) {
      config.setDeviceId(deviceId);
    }
    await Countly.initWithConfig(config);

    final deviceIdMigrated = await userRepository.isCountlyDeviceIdMigrated();
    if (!deviceIdMigrated && deviceId != null) {
      await Countly.instance.deviceId.changeWithMerge(deviceId);
      await userRepository.markCountlyDeviceIdMigrated();
    }

    await Countly.askForNotificationPermission();
    await Countly.appLoadingFinished();
```

> From your description, it sounds like the Countly.askForNotificationPermission method is not returning as expected, resulting in a "never-ending" call. Is this correct? 

Yes, it's correct.

> If possible, could you also specify any logs or error messages you’re seeing in the console?

I can't see anything relevant.
But here it is 

> V/SurfaceView@aa72c3(26236): Layout: x=0 y=0 w=1080 h=2358, frame=Rect(0, 0 - 1080, 2358)
> I/ViewRootImpl@834a70f[MainActivity](26236): destroyHardwareResources: Callers=android.view.ViewRootImpl.performTraversals:3932 android.view.ViewRootImpl.doTraversal:3288 android.view.ViewRootImpl$TraversalRunnable.run:11344 android.view.Choreographer$CallbackRecord.run:1689 android.view.Choreographer$CallbackRecord.run:1698 android.view.Choreographer.doCallbacks:1153 android.view.Choreographer.doFrame:1079 android.view.Choreographer$FrameDisplayEventReceiver.run:1646 android.os.Handler.handleCallback:958 android.os.Handler.dispatchMessage:99 
> D/OpenGLRenderer(26236): CacheManager::trimMemory(20)
> I/ViewRootImpl@834a70f[MainActivity](26236): Relayout returned: old=(0,0,1080,2400) new=(0,0,1080,2400) relayoutAsync=false req=(1080,2400)8 dur=19 res=0x402 s={false 0x0} ch=true seqId=0
> I/SurfaceView@aa72c3(26236): windowStopped(true) false io.flutter.embedding.android.FlutterSurfaceView{aa72c3 V.E...... ......ID 0,0-1080,2358} of ViewRootImpl@834a70f[MainActivity]
> D/SurfaceView@aa72c3(26236): updateSurface: surface is not valid
> I/SurfaceView@aa72c3(26236): releaseSurfaces: viewRoot = ViewRootImpl@834a70f[MainActivity]
> D/SurfaceView@aa72c3(26236): updateSurface: surface is not valid
> I/SurfaceView@aa72c3(26236): releaseSurfaces: viewRoot = ViewRootImpl@834a70f[MainActivity]
> I/ViewRootImpl@834a70f[MainActivity](26236): stopped(true) old = false
> D/ViewRootImpl@834a70f[MainActivity](26236): WindowStopped on com.my.app.id/com.my.app.id.app.MainActivity set to true
> D/OpenGLRenderer(26236): CacheManager::trimMemory(20)
> D/SurfaceView@aa72c3(26236): updateSurface: surface is not valid
> I/SurfaceView@aa72c3(26236): releaseSurfaces: viewRoot = ViewRootImpl@834a70f[MainActivity]
> D/SurfaceView@aa72c3(26236): updateSurface: surface is not valid
> I/SurfaceView@aa72c3(26236): releaseSurfaces: viewRoot = ViewRootImpl@834a70f[MainActivity]
> I/ViewRootImpl(26236): updatePointerIcon pointerType = 1, calling pid = 26236
> D/InputManagerGlobal(26236): setPointerIconType iconId = 1, callingPid = 26236
> I/ViewRootImpl@834a70f[MainActivity](26236): onDisplayChanged oldDisplayState=2 newDisplayState=1
> I/ViewRootImpl@834a70f[MainActivity](26236): onDisplayChanged oldDisplayState=1 newDisplayState=1

**Comment by ijunaid** [comment]
Thank you for your response. Your code looks fine.

Could you please share the Countly SDK logs? Here’s a link that explains how to obtain the Android SDK logs:
https://support.countly.com/hc/en-us/articles/900000908046-Getting-Started-with-SDKs#h_01HABSX9KXS2ND6YQKFVC1PRWZ

**Comment by sbatezat** [comment]
> Thank you for your response. Your code looks fine.
> 
> Could you please share the Countly SDK logs? Here’s a link that explains how to obtain the Android SDK logs: https://support.countly.com/hc/en-us/articles/900000908046-Getting-Started-with-SDKs#h_01HABSX9KXS2ND6YQKFVC1PRWZ

My bad, logging was disabled.
Here it is

```
2024-10-29 09:25:03.343 30015-30101 flutter                 com.my.app.id.myapp  I  [CountlyFlutter] DEBUG: Calling "initWithConfig"
2024-10-29 09:25:03.362 30015-30101 flutter                 com.my.app.id.myapp  I  [CountlyFlutter] INFO: "_configToJson", Device ID provided: [HIDDEN]
2024-10-29 09:25:03.365 30015-30101 flutter                 com.my.app.id.myapp  I  [CountlyFlutter] INFO: "_configToJson", value provided for loggingEnabled: [true]
2024-10-29 09:25:03.366 30015-30101 flutter                 com.my.app.id.myapp  I  [CountlyFlutter] INFO: "_configToJson", value provided for enableUnhandledCrashReporting: [true]
2024-10-29 09:25:03.370 30015-30101 flutter                 com.my.app.id.myapp  I  [CountlyFlutter] INFO: "_configToJson", value provided for trackAppStartTime: [true]
2024-10-29 09:25:03.371 30015-30101 flutter                 com.my.app.id.myapp  I  [CountlyFlutter] INFO: "_configToJson", value provided for enableForegroundBackground: [true]
2024-10-29 09:25:03.371 30015-30101 flutter                 com.my.app.id.myapp  I  [CountlyFlutter] INFO: "_configToJson", value provided for enableManualAppLoaded: [true]
2024-10-29 09:25:03.374 30015-30101 flutter                 com.my.app.id.myapp  I  [CountlyFlutter] INFO: "_configToJson", value provided for remoteConfigAutomaticTriggers: [false]
2024-10-29 09:25:03.375 30015-30101 flutter                 com.my.app.id.myapp  I  [CountlyFlutter] INFO: "_configToJson", value provided for remoteConfigValueCaching: [false]
2024-10-29 09:25:03.383 30015-30015 Countly                 com.my.app.id.myapp  D  Enabling logging
2024-10-29 09:25:03.384 30015-30015 Countly                 com.my.app.id.myapp  D  [Init] Initializing Countly [dart-flutterb-android] SDK version [24.7.1] default name[java-native-android] default version[24.7.1]
2024-10-29 09:25:03.384 30015-30015 Countly                 com.my.app.id.myapp  D  [Init] Using explicitly provided context
2024-10-29 09:25:03.384 30015-30015 Countly                 com.my.app.id.myapp  V  [Init] Removing trailing '/' from provided server url
2024-10-29 09:25:03.384 30015-30015 Countly                 com.my.app.id.myapp  D  [Init] SDK initialised with the URL:[HIDDEN] and the appKey:[HIDDEN]
2024-10-29 09:25:03.384 30015-30015 Countly                 com.my.app.id.myapp  I  [Init] Checking init parameters
2024-10-29 09:25:03.388 30015-30015 Countly                 com.my.app.id.myapp  I  [Init] Provided Context [Application], it's superclass: [ContextWrapper]
2024-10-29 09:25:03.389 30015-30015 Countly                 com.my.app.id.myapp  I  [Init] Getting in the 'else' block
2024-10-29 09:25:12.906 30015-3062  Countly                 com.my.app.id.myapp  V  [onTimer] Calling heartbeat, Activity count:[0]
2024-10-29 09:25:12.906 30015-3062  Countly                 com.my.app.id.myapp  V  [ModuleRequestQueue] forceSendingEvents, forced:[true], event count:[0]
2024-10-29 09:25:12.906 30015-3062  Countly                 com.my.app.id.myapp  D  [ModuleUserProfile] saveInternal
2024-10-29 09:25:12.907 30015-3062  Countly                 com.my.app.id.myapp  D  [ModuleUserProfile] saveInternal, no user data to save
2024-10-29 09:25:12.907 30015-3062  Countly                 com.my.app.id.myapp  V  [ConnectionQueue] tick, IsRQEmpty:[true], HasOngoingProcess:[false], OngoingProcess_Done:[true]
2024-10-29 09:25:17.688 30015-30101 flutter                 com.my.app.id.myapp  I  [CountlyFlutter] DEBUG: Calling "askForNotificationPermission"
2024-10-29 09:25:17.693 30015-30015 CountlyFlutterPlugin    com.my.app.id.myapp  I  Method name: askForNotificationPermission
2024-10-29 09:25:17.693 30015-30015 CountlyFlutterPlugin    com.my.app.id.myapp  I  Method arguments: []
2024-10-29 09:25:17.697 30015-30015 Countly                 com.my.app.id.myapp  I  [CountlyPush, init] Initializing Countly Push, App:[true], provider:[null]
2024-10-29 09:25:17.707 30015-30015 Countly                 com.my.app.id.myapp  D  [ModuleCrash] Uncaught crash handler triggered
2024-10-29 09:25:17.719 30015-30015 Countly                 com.my.app.id.myapp  W  [ModuleCrash] prepareStackTrace, Value length exceeds limit of 200 characters. Truncating value to 200 characters. Truncated to Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.iid.FirebaseInstanceId" on path: DexPathList[[zip file "/data/app/~~lXKpw0hOza72pQyRbdw-iA==/com.my.app.id
2024-10-29 09:25:17.737 30015-30015 Countly                 com.my.app.id.myapp  D  [ModuleCrash] Calling crashFilterCheck
2024-10-29 09:25:17.737 30015-30015 Countly                 com.my.app.id.myapp  D  [ModuleCrash] sendCrashReportToQueue
2024-10-29 09:25:17.750 30015-30015 Countly                 com.my.app.id.myapp  D  [Connection Queue] sendCrashReport
2024-10-29 09:25:17.774 30015-30015 Countly                 com.my.app.id.myapp  V  [CountlyStore] addRequest, s:[true] new q size:[1] r:[app_key=HIDDEN&device_id=HIDDEN&timestamp=1730190317750&sdk_version=24.7.1&sdk_name=dart-flutterb-android&av=1.7.0&hour=9&dow=2&tz=60&crash=%7B%22_run%22%3A%22125%22%2C%22_os%22%3A%22Android%22%2C%22_ram_total%22%3A%225371%22%2C%22_os_version%22%3A%2214%22%2C%22_cpu%22%3A%22arm64-v8a%22%2C%22_online%22%3A%22true%22%2C%22_error%22%3A%22java.lang.NoClassDefFoundError%3A+Failed+resolution+of%3A+Lcom%5C%2Fgoogle%5C%2Ffirebase%5C%2Fiid%5C%2FFirebaseInstanceId%3B%5Cn%5Ctat+ly.count.dart.countly_flutter.CountlyFlutterPlugin.onMethodCall%28CountlyFlutterPlugin.java%3A443%29%5Cn%5Ctat+io.flutter.plugin.common.MethodChannel%24IncomingMethodCallHandler.onMessage%28MethodChannel.java%3A267%29%5Cn%5Ctat+io.flutter.embedding.engine.dart.DartMessenger.invokeHandler%28DartMessenger.java%3A292%29%5Cn%5Ctat+io.flutter.embedding.engine.dart.DartMessenger.lambda%24dispatchMessageToQueue%240%24io-flutter-embedding-engine-dart-DartMessenger%28DartMessenger.java%3A319%29%5Cn%5Ctat+io.flutter.embedding.engine.dart.DartMessenger%24%24ExternalSyntheticLambda0.run%28Unknown+Source%3A12%29%5Cn%5Ctat+android.os.Handler.handleCallback%28Handler.java%3A958%29%5Cn%5Ctat+android.os.Handler.dispatchMessage%28Handler.java%3A99%29%5Cn%5Ctat+android.os.Looper.loopOnce%28Looper.java%3A230%29%5Cn%5Ctat+android.os.Looper.loop%28Looper.java%3A319%29%5Cn%5Ctat+android.app.ActivityThread.main%28ActivityThread.java%3A8919%29%5Cn%5Ctat+java.lang.reflect.Method.invoke%28Native+Method%29%5Cn%5Ctat+com.android.internal.os.RuntimeInit%24MethodAndArgsCaller.run%28RuntimeInit.java%3A578%29%5Cn%5Ctat+com.android.internal.os.ZygoteInit.main%28ZygoteInit.java%3A1103%29%5CnCaused+by%3A+java.lang.ClassNotFoundException%3A+Didn%27t+find+class+%5C%22com.google.firebase.iid.FirebaseInstanceId%5C%22+on+path%3A+DexPathList%5B%5Bzip+file+%5C%22%5C%2Fdata%5C%2Fapp%5C%2F%7E%7ElXKpw0hOza72pQyRbdw-iA%3D%3D%5C%2Fcom.my.app.idr%5Cn%5Ctat+dalvik.system.BaseDexClassLoader.findClass%28BaseDexClassLoader.java%3A259%29%5Cn%5Ctat+java.lang.ClassLoader.loadClass%28ClassLoader.java%3A379%29%5Cn%5Ctat+java.lang.ClassLoader.loadClass%28ClassLoader.java%3A312%29%5Cn%5Ct...+13+more%22%2C%22_nonfatal%22%3A%22false%22%2C%22_manufacturer%22%3A%22samsung%22%2C%22_device%22%3A%22SM-A528B%22%2C%22_orientation%22%3A%22Portrait%22%2C%22_ram_current%22%3A%223736%22%2C%22_app_version%22%3A%221.7.0%22%2C%22_root%22%3A%22false%22%2C%22_disk_current%22%3A%226633%22%2C%22_has_hinge%22%3A%22false%22%2C%22_ob%22%3A0%2C%22_resolution%22%3A%221080x2270%22%2C%22_disk_total%22%3A%226644%22%2C%22_bat%22%3A%2276.0%22%2C%22_background%22%3A%22true%22%2C%22_opengl%22%3A%223%22%2C%22_muted%22%3A%22false%22%7D]
2024-10-29 09:25:17.786 30015-30015 Countly                 com.my.app.id.myapp  V  [ConnectionQueue] tick, IsRQEmpty:[false], HasOngoingProcess:[false], OngoingProcess_Done:[true]
2024-10-29 09:25:17.786 30015-30015 Countly                 com.my.app.id.myapp  D  [ConnectionQueue] tick, Starting ConnectionProcessor
2024-10-29 09:25:17.787 30015-6448  Countly                 com.my.app.id.myapp  I  [Connection Processor] Starting to run, there are [1] requests stored
2024-10-29 09:25:17.787 30015-6448  Countly                 com.my.app.id.myapp  I  [ConnectionProcessor] Checking if the request is older than:[0] hours
2024-10-29 09:25:17.787 30015-6448  Countly                 com.my.app.id.myapp  V  [ConnectionProcessor] isRequestTooOld, No request drop age set. Request will bypass age checks
2024-10-29 09:25:17.793 30015-6448  Countly                 com.my.app.id.myapp  V  [RequestQueue] Calling 'isDeviceAppCrawler'
2024-10-29 09:25:17.797 30015-6448  Countly                 com.my.app.id.myapp  V  [ConnectionProcessor] The following checksum was added:[0998bed7e0b7c8e4e255a213723749be952a4a615415f188f0c78483eba4b93f]
2024-10-29 09:25:17.798 30015-6448  Countly                 com.my.app.id.myapp  V  [ConnectionProcessor] Has picturePath [false]
2024-10-29 09:25:19.392 30015-6448  Countly                 com.my.app.id.myapp  V  [RequestQueue] Calling 'isHttpPostForced'
2024-10-29 09:25:19.393 30015-6448  Countly                 com.my.app.id.myapp  V  [ConnectionProcessor] Using HTTP POST: [true] forced:[false] length:[true] crash:[true] | Approx data size: [3910 B]
2024-10-29 09:25:19.393 30015-6448  Countly                 com.my.app.id.myapp  D  [ConnectionProcessor] run, TIMING Setup server request took:[1599.125676] ms
2024-10-29 09:25:19.403 30015-6448  Countly                 com.my.app.id.myapp  D  [Connection Processor] code:[200], response:[{"result":"Success"}], response size:[20 B], request: app_key=HIDDEN&device_id=HIDDEN&timestamp=1730190317750&sdk_version=24.7.1&sdk_name=dart-flutterb-android&av=1.7.0&hour=9&dow=2&tz=60&crash=%7B%22_run%22%3A%22125%22%2C%22_os%22%3A%22Android%22%2C%22_ram_total%22%3A%225371%22%2C%22_os_version%22%3A%2214%22%2C%22_cpu%22%3A%22arm64-v8a%22%2C%22_online%22%3A%22true%22%2C%22_error%22%3A%22java.lang.NoClassDefFoundError%3A+Failed+resolution+of%3A+Lcom%5C%2Fgoogle%5C%2Ffirebase%5C%2Fiid%5C%2FFirebaseInstanceId%3B%5Cn%5Ctat+ly.count.dart.countly_flutter.CountlyFlutterPlugin.onMethodCall%28CountlyFlutterPlugin.java%3A443%29%5Cn%5Ctat+io.flutter.plugin.common.MethodChannel%24IncomingMethodCallHandler.onMessage%28MethodChannel.java%3A267%29%5Cn%5Ctat+io.flutter.embedding.engine.dart.DartMessenger.invokeHandler%28DartMessenger.java%3A292%29%5Cn%5Ctat+io.flutter.embedding.engine.dart.DartMessenger.lambda%24dispatchMessageToQueue%240%24io-flutter-embedding-engine-dart-DartMessenger%28DartMessenger.java%3A319%29%5Cn%5Ctat+io.flutter.embedding.engine.dart.DartMessenger%24%24ExternalSyntheticLambda0.run%28Unknown+Source%3A12%29%5Cn%5Ctat+android.os.Handler.handleCallback%28Handler.java%3A958%29%5Cn%5Ctat+android.os.Handler.dispatchMessage%28Handler.java%3A99%29%5Cn%5Ctat+android.os.Looper.loopOnce%28Looper.java%3A230%29%5Cn%5Ctat+android.os.Looper.loop%28Looper.java%3A319%29%5Cn%5Ctat+android.app.ActivityThread.main%28ActivityThread.java%3A8919%29%5Cn%5Ctat+java.lang.reflect.Method.invoke%28Native+Method%29%5Cn%5Ctat+com.android.internal.os.RuntimeInit%24MethodAndArgsCaller.run%28RuntimeInit.java%3A578%29%5Cn%5Ctat+com.android.internal.os.ZygoteInit.main%28ZygoteInit.java%3A1103%29%5CnCaused+by%3A+java.lang.ClassNotFoundException%3A+Didn%27t+find+class+%5C%22com.google.firebase.iid.FirebaseInstanceId%5C%22+on+path%3A+DexPathList%5B%5Bzip+file+%5C%22%5C%2Fdata%5C%2Fapp%5C%2F%7E%7ElXKpw0hOza72pQyRbdw-iA%3D%3D%5C%[2Fcom.my.app.id](http://2fcom.my.app.id/)%5Cn%5Ctat+dalvik.system.BaseDexClassLoader.findClass%28BaseDexClassLoader.java%3A259%29%5Cn%5Ctat+java.lang.ClassLoader.loadClass%28ClassLoader.java%3A379%29%5Cn%5Ctat+java.lang.ClassLoader.loadClass%28ClassLoader.java%3A312%29%5Cn%5Ct...+13+more%22%2C%22_nonfatal%22%3A%22false%22%2C%22_manufacturer%22%3A%22samsung%22%2C%22_device%22%3A%22SM-A528B%22%2C%22_orientation%22%3A%22Portrait%22%2C%22_ram_current%22%3A%223736%22%2C%22_app_version%22%3A%221.7.0%22%2C%22_root%22%3A%22false%22%2C%22_disk_current%22%3A%226633%22%2C%22_has_hinge%22%3A%22false%22%2C%22_ob%22%3A0%2C%22_resolution%22%3A%221080x2270%22%2C%22_disk_total%22%3A%226644%22%2C%22_bat%22%3A%2276.0%22%2C%22_background%22%3A%22true%22%2C%22_opengl%22%3A%223%22%2C%22_muted%22%3A%22false%22%7D&rr=0, url: HIDDEN, Reading stream took:[9.715365] ms
2024-10-29 09:25:19.404 30015-6448  Countly                 com.my.app.id.myapp  V  [ConnectionProcessor] Response was a success
2024-10-29 09:25:19.406 30015-6448  Countly                 com.my.app.id.myapp  D  [ConnectionProcessor] run, TIMING Handling response took:[13.073594] ms
2024-10-29 09:25:19.407 30015-6448  Countly                 com.my.app.id.myapp  V  [Connection Processor] Starting to run, there are [0] requests stored
2024-10-29 09:25:19.407 30015-6448  Countly                 com.my.app.id.myapp  I  [ConnectionProcessor] No requests in the queue, request queue skipped
2024-10-29 09:25:19.407 30015-6448  Countly                 com.my.app.id.myapp  V  [ConnectionProcessor] run, TIMING Whole queue took:[1619.917864] ms
```

**Comment by ijunaid** [comment]
Thank you for sharing the logs! It seems the initialization is set up correctly. However, there’s an error related to `FirebaseInstanceId`, which may indicate that Firebase dependencies need to be added in the Gradle configuration.

Could you please verify the push notification setup using this link:
https://support.countly.com/hc/en-us/articles/360037944212-Flutter#h_01H930GAQ6C3B3RYSEXZX4ZY3F

**Comment by sbatezat** [comment]
Everything seems ok, I can't proove it for now, but I've got the feeling there is conflict with my other dependencies :

```
firebase_core: ^3.6.0
  firebase_messaging: ^15.1.3
```

Indeed, my app is already using FCM for server side notification on business use case.
My app is so using the latest FCM SDK (24.0.1).

I see java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/firebase/iid/FirebaseInstanceId;

Is it possible Google has removed this class on its latest SDK and you'r still using it?

**Comment by sbatezat** [comment]
I can confirm it's a conflict between "my" firebase version and "yours"
Indeed, if I use the "update-fcm" branch from https://github.com/Countly/countly-sdk-flutter-bridge/pull/276, the app is not blocked anymore.

**Comment by sbatezat** [comment]
> it still needs some additional changes for full backward compatibility. This is a breaking change, and we plan to release it in a future update.

Could you please elaborate on this? What are the risks to use that branch? I need notifications on my app next release and unfortunately i can't wait... i see two workaround: use your branch, but I need more info on it, or downgrade the SDK I used (with possibly lot of issues...)

**Comment by ijunaid** [comment]
If someone is using an older Firebase version, they may encounter issues with the updated implementation. We’re working to ensure compatibility with these older versions while avoiding any breaking changes. We plan to release this update along with other changes at the start of next week.

**Comment by sbatezat** [comment]
fixed in 24.11.0
